### ISSUE TITLE
Add VAT/Tax field to Buyer Information on invoice

### ISSUE THREAD

**Issue by nodecheck** [btcpayserver/btcpayserver#614]
Would it be possible to add to the Buyer Information section for invoices a VAT ID/Tax ID field?  This would be ideal for people wanting to process company transactions where completing such information would be a requisite.

**Comment by NicolasDorier** [comment]
ping @JeffVandrewJr any idea what can be done which would be perfect?

**Comment by JeffVandrewJr** [comment]
Right now BTCPay optionally allows input of a `taxIncluded` field to separate collected tax from revenue. It doesn't currently separate different types of tax, the idea being that you should have separate sales tax software for that. In the USA, we don't have VAT, but for sales tax we have special vendors that specialize in sales tax software. 

If you wanted to drastically expand the POS app functionality, you could theoretically allow the user to input different tax names and rates and have the POS app automatically calculate it. For instance, you could have the user create a tax called "New Jersey Sales Tax" with a rate of 6.625%, and the user could select which items were subject to that tax. The user would need to be responsible for updating any rates over time, as BTCPay can't monitor tax rates in thousands of jurisdictions.

**Comment by Kukks** [comment]
With that amount of data, it would make more sense to add it to `posData` when submitting it

**Comment by NicolasDorier** [comment]
@Kukks yeah I think that is good idea.

**Comment by nodecheck** [comment]
From our perspective in the EU, it's much simpler, since for example, the UK has VAT/Tax of 20% so it means each transaction would have this by default.  For other countries with more complex tax whereby they have differing levels, that's something completely different and difficult to manage obviously.

For me, I want to set the rate to 20%, but I'd like the customer to be able to put in during checkout, their VAT/Tax ID, so that it appears in the invoice.  Of course, their address should be able to appear there also on the invoice.  As at present from the PoS shop this doesn't appear on the generated invoice.

**Comment by second2falcon** [comment]
Is there any update/change on this? Would be nice to see something where we might be able to enter the VAT for specific items (in the Netherlands, depending on the kind of product you pay different tax). Maybe in raw editor for the items in a PoS app.

**Comment by Zoz92** [comment]
Can add tax on the receipt and détails of the card ?

**Comment by pavlenex** [comment]
@Zoz92 Yes, we're working on this for 1.9.0 perhaps.

**Comment by pavlenex** [comment]
@Zoz92 and @second2falcon You guys use our Point of sale app, and would like to have a tax field there as well, or is this about the API and custom integrations?

**Comment by pavlenex** [comment]
> From our perspective in the EU, it's much simpler, since for example, the UK has VAT/Tax of 20% so it means each transaction would have this by default. For other countries with more complex tax whereby they have differing levels, that's something completely different and difficult to manage obviously.
> 
> For me, I want to set the rate to 20%, but I'd like the customer to be able to put in during checkout, their VAT/Tax ID, so that it appears in the invoice. Of course, their address should be able to appear there also on the invoice. As at present from the PoS shop this doesn't appear on the generated invoice.

@nodecheck We've added [custom form builders in 1.8.0](https://blog.btcpayserver.org/btcpay-server-1-8-0/#-custom-checkout-forms) and it would allow you to cater this specific usecase of having a field before checkout where customers can input their Vat ID.

**Comment by Zoz92** [comment]
i am not a tech guy but in france for meal/wine, whe have différent tax 10% and 20%, people need details on receipt of what they eat and they need the final tax.
I don't use api atm, i tryed to do some GET request but i am not a dev so it is complicate for me 
I found the (BTCPAY_INSTANCE + "/api/v1/stores/" + storeId + "/invoices/" + invoiceId, 'GET', auth) but i am struggle with this auth system maybe a secret would be great 
The  best would be to have the automatic send of the receip to the buyer and a automatic webhook receip

Sorry for my poor english, if you need more explain ask me !

**Comment by pavlenex** [comment]
@Zoz92 No worries, most of btcpay contributors including myself aren't native speakers, so don't bother about that.

Could you explain to me how you use btcpay currently and what kind of business it is? Do you use point of sale app or something else? We can definitely expose tax in the receipt, i'm just trying to learn about specific use-cases so we can find optimal solution.

**Comment by Zoz92** [comment]
I am running some btcpayserver node for knowing how it is working, i am part of bitcoin family try to orange pill some little business, clearly some of them just use the 20% tax so it is not difficult for them with the export but client also need this on the receipt.
Exemple
https://france3-regions.francetvinfo.fr/image/idtYNQXl3eDmq4eRAbHHJCIyo9I/600x400/regions/2020/06/09/5edeb52a4013b_addition.jpg

**Comment by Zoz92** [comment]
I will looks this custom forms on the 1.8

**Comment by second2falcon** [comment]
> @Zoz92 and @second2falcon You guys use our Point of sale app, and would like to have a tax field there as well, or is this about the API and custom integrations?

I tried to use it as a proof of concept, but in the end, that was one of the reasons it didn't end well. 
I tried using the invoice in the web app.

**Comment by NicolasDorier** [comment]
I think we should allow people to show whatever information from the Metadata in the receipt.
Today you are asking VAT/Tax tomorrow it will be other fields, so we need to make that configurable.

**Comment by Kukks** [comment]
It is configurable through the receiptData option

**Comment by dennisreimann** [comment]
Which we have to make available on the UI or display it by default. I think this is part of #4716.

**Comment by Kukks** [comment]
See #4774

**Comment by Kukks** [comment]
So to sum up the work so far:
* We now show all metadata in the invoice summary.
* The receipt page shows the cart, tip and discount from a POS app
*  A form is able to be configured to
  * have inputs that write data to the receipt page directly
  * have configuration to copy data to the receipt page

**Comment by pavlenex** [comment]
Thanks for the summary @Kukks so what you're saying is if there's a vat field, it will be shown to the invoice and on the receipt? Or are some further tweaks needed to show it on a receipt as well? If it's the first one, then imo this resolves the issue.

**Comment by Zoz92** [comment]
Hello, looks amazing, does any plan to add the VAT/Tax field on the receipt/invoice ?
And maybe some PRINT button to export in one page the invoice (actually if i do print, that do 6 pages a4)

**Comment by Kukks** [comment]
> Thanks for the summary @Kukks so what you're saying is if there's a vat field, it will be shown to the invoice and on the receipt? Or are some further tweaks needed to show it on a receipt as well? If it's the first one, then imo this resolves the issue.

Depends on how this VAT field would work.. Is it something on the item, is it computed from the total, etc 

Most likely it won't show. But a lot of parts have been added to make this close to possible. Once #4792  is merged, we can store arbitrary data on each item in the pos app, which allows for example storing a "Tax amount". and then we can probably customize how data is shown on the receipt to be a bit more dynamic

**Comment by second2falcon** [comment]
> > Thanks for the summary @Kukks so what you're saying is if there's a vat field, it will be shown to the invoice and on the receipt? Or are some further tweaks needed to show it on a receipt as well? If it's the first one, then imo this resolves the issue.
> 
> 
> 
> Depends on how this VAT field would work.. Is it something on the item, is it computed from the total, etc 
> 
> 
> 
> Most likely it won't show. But a lot of parts have been added to make this close to possible. Once #4792  is merged, we can store arbitrary data on each item in the pos app, which allows for example storing a "Tax amount". and then we can probably customize how data is shown on the receipt to be a bit more dynamic

I mean, theoretically both per item tax and a total tax would work. 
Which one to choose depends more on the products being sold. 
So if items have different tax rates, a total tax wouldn't work. 

(Anyone correct me if I'm wrong)

**Comment by pavlenex** [comment]
Feels like what @Kukks describes lays out the foundation for the later work and we can see how to properly expose the tax field. In that case I'd leave this issue opened, move it out of the 1.9.0  and we can scope it for 1.9.10.

**Comment by ndeet** [comment]
A bit of context from the devcall. If we want to have full support there are many things to consider:
- in most EU countries, you have 2 tax rates, a reduced one for food and a normal for other goods, e.g. in Austria 10% and 20%, Germany 7% and 19% etc.
- beyond the EU, Canada and other countries have different rules
- edge cases like is a tip taxable or not differs from country to country etc.
- tax configurable by line item
- on cart / receipt you need to group them per tax
- 0% reverse charge tax in b2b case in the EU

In PHP land there is [one library ](https://github.com/commerceguys/tax) that tries to cover most cases, not only the EU. There should be something out there also for C#?

**Comment by me639276639936** [comment]
Yes please - option to add tax rates. USA has state, city, county, hospitality taxes, and more. And if we are serious about using this as a POS, all these lines items need to be there for accounting purpose.

**Comment by Marshamolly** [comment]
Idk what to do to get my biconical minning deposit in one of my accounts

**Comment by me639276639936** [comment]
This is not a BTCpay server question. You need to go learn about mining.

**Comment by me639276639936** [comment]
> Idk what to do to get my biconical minning deposit in one of my accounts

This is not a mining forum. This is specific to BTCPay server issues.

**Comment by Zoz92** [comment]
@pavlenex any new roadmap about the feature ?

**Comment by Huckelbarry** [comment]
No. You?

**Comment by pavlenex** [comment]
@Zoz92 it should be done in the next release check #5077

**Comment by Zoz92** [comment]
Thanks @pavlenex

**Comment by pavlenex** [comment]
This one was done as part of https://github.com/btcpayserver/btcpayserver/pull/5077

**Comment by dennisreimann** [comment]
> This one was done as part of https://github.com/btcpayserver/btcpayserver/pull/5077

@pavlenex was it? Or is it just that the additional info is now also part of the receipt?

**Comment by dennisreimann** [comment]
tbh I'm not sure and there's so much back and forth here that it's not even clear what the intended goal was supposed to be.

**Comment by Zoz92** [comment]
I think you could take the comments has the intended goals:
https://github.com/btcpayserver/btcpayserver/issues/614#issuecomment-1514850176

**Comment by pavlenex** [comment]
@Zoz92 you're using our point of sale app, correct?

**Comment by Zoz92** [comment]
Just for testing only

**Comment by pavlenex** [comment]
I am trying to understand if your use-case is through point-of sale app or not, or are you using different parts of our software (API, custom integration) to achieve your goal.

**Comment by dennisreimann** [comment]
We recently discussed this in the dev meeting and consensus is to keep the POS lean. Things like tax per lineityem are beyond the use case of the POS and a good indicater, that you'd rather want to use a more advanced solution: Please look at our ecommerce integrations like WooCOmmerce or Magento.

**Comment by pavlenex** [comment]
No, I don't agree with that. Keeping lean if people use it as an e-commerce shop, but if this is the use-case for retail, which sounds like it is, then it's a core feature.

**Comment by Zoz92** [comment]
No like i said i am not a tech guy https://github.com/btcpayserver/btcpayserver/issues/614#issuecomment-1456500570
If you want this app should work simple for commerce why you should use a woocommerce ?

**Comment by pavlenex** [comment]
I understand, but that's out of scope, point of sale is mean for retail, not to be an e-commerce solution, hence we don't cater for online shop requests as @dennisreimann mentioned.

**Comment by Zoz92** [comment]
As said before in retail, seller and buyer need the vat / tax field on the receipt
https://github.com/btcpayserver/btcpayserver/issues/614#issuecomment-1514850176
https://github.com/btcpayserver/btcpayserver/issues/614#issuecomment-1574524915
https://github.com/btcpayserver/btcpayserver/issues/614#issuecomment-1456526164

**Comment by Zoz92** [comment]
> We recently discussed this in the dev meeting and consensus is to keep the POS lean. Things like tax per lineitem are beyond the use case of the POS and a good indicator, that you'd rather want to use a more advanced solution: Please look at our ecommerce integrations like WooCOmmerce or Magento.

Retail commerce like restaurant use differents tax , it is not a e-commerce solution, it is a common store in the world

**Comment by me639276639936** [comment]
All of our receipts in the US given by POS have at least one final line item for tax.Perhaps this is a foreign concept to you all, but not to us. :)Sent from my iPhoneOn Jul 11, 2023, at 6:37 AM, Zoz ***@***.***> wrote:﻿

We recently discussed this in the dev meeting and consensus is to keep the POS lean. Things like tax per lineitem are beyond the use case of the POS and a good indicator, that you'd rather want to use a more advanced solution: Please look at our ecommerce integrations like WooCOmmerce or Magento.

Retail commerce like restaurant use differents tax , it is not a e-commerce solution, it is a common store in the world

**Comment by Kukks** [comment]
> No, I don't agree with that. Keeping lean if people use it as an e-commerce shop, but if this is the use-case for retail, which sounds like it is, then it's a core feature.

This shouldn't be a core feature. The worldwide differences around taxes and requirements mean this feature must be incredibly complex. My suggestion is to do a separate plugin where we can iterate on extending outside of main releases.
